fix(repo setup): Skip aliases for unloaded toolchains

kilian-funk opened the pull request:
Some platforms don't contain every version, e.g. s390x doesn't have 3.8, which is
indicated by a missing sha256 value. When this happens, no repository for the runtime
is created (`python_repository` rule).

Similar logic needs to be in the toolchains setup logic because otherwise a reference
to an undefined repository exists in the select() expression of the aliases. Because those
references are lazily evaluated, they don't always cause a problem, but do mean that
query operations (e.g., `rdeps()`) can't work and the order of entries is important
(which is surprising).

Closes #1472

rickeylev commented on python/private/toolchains_repo.bzl:
```diff
@@ -195,6 +196,9 @@ load("{rules_python}//python:pip.bzl", _compile_pip_requirements = "compile_pip_
 
 host_platform = "{host_platform}"
 interpreter = "@{py_repository}_{host_platform}//:{python3_binary_path}"
+PLATFORMS = [
```
Please take this out of defs.bzl -- defs.bzl is a public entry point, and this isn't a detail we want to make part of our public API.

Instead, it can be inlined in build_contents directly:
```
_plaforms = [
  {loaded_platforms}
]
```

kilian-funk replied:
IMO the selection of platforms is valuable information, but I am not sure if this is the right form to make that available. So, I am fine with removing it.

kilian-funk replied:
Changed as suggested.

rickeylev commented on CHANGELOG.md:
```diff
@@ -19,6 +19,13 @@ A brief description of the categories of changes:
 
+### Fixed
+
+* Skip aliases for unloaded toolchains. Toolchains without a checksum are skipped
+  in the repo load phase. These skipped repos now don't have aliases in the toolchains repo.
+  This is relevant when working with older python versions that don't have full platform 
+  support.
```
```suggestion
* Skip aliases for unloaded toolchains. Some Python versions that don't have full
  platform support, and referencing their undefined repositories can break operations
  like `bazel query rdeps(...)`.
```

kilian-funk replied:
Done.

rickeylev reviewed (changes requested):
mostly lgtm, just move the platforms constant out of the generated defs.bzl

rickeylev commented:
CI is unhappy, but it's not immediately apparent why. It gives me the feeling there might be a syntax error somewhere? Or maybe what's happening is the BUILD file is trying to load defs.bzl before something defs.bzl needs is ready? Maybe removing that `load(defs.bzl)` as I commented will fix it.

kilian-funk commented:
@rickeylev I have addressed your concerns. CI is happy now, even without knowing the root cause of the original failures.

Could you please re-review?

aignas reviewed (approved):
LGTM
